File: 2018/Python/day19.py
from collections import deque

FUNCS = {'addr': lambda xs, x, y: xs[x] + xs[y],
         'addi': lambda xs, x, y: xs[x] + y,
         'mulr': lambda xs, x, y: xs[x] * xs[y],
         'muli': lambda xs, x, y: xs[x] * y,
         'banr': lambda xs, x, y: xs[x] & xs[y],
         'bani': lambda xs, x, y: xs[x] & y,
         'borr': lambda xs, x, y: xs[x] | xs[y],
         'bori': lambda xs, x, y: xs[x] | y,
         'setr': lambda xs, x, y: xs[x],
         'seti': lambda xs, x, y: x,
         'gtir': lambda xs, x, y: 1 if x > xs[y] else 0,
         'gtri': lambda xs, x, y: 1 if xs[x] > y else 0,
         'gtrr': lambda xs, x, y: 1 if xs[x] > xs[y] else 0,
         'eqir': lambda xs, x, y: 1 if x == xs[y] else 0,
         'eqri': lambda xs, x, y: 1 if xs[x] == y else 0,
         'eqrr': lambda xs, x, y: 1 if xs[x] == xs[y] else 0}

def read_file(filename):
    with open(filename) as f:
        pointer = int(f.readline().split()[1])
        lines = f.read().splitlines()
    return {i: line for i, line in enumerate(lines)}, pointer

def call(regs, data, inst, debug=False):
    f, r1, r2, res = data[regs[inst]].split()
    regs[int(res)] = FUNCS[f](regs, int(r1), int(r2))
    # print(f'{regs} after {f}')
    regs[inst] += 1

def run_code(data, inst, debug=False):
    regs = [1, 0, 0, 0, 0 ,0]
    previous = set()
    count = 0
    while regs[inst] in data:
        call(regs, data, inst, debug=debug)
    print(f'Register 0: {regs[0]}')


def main(debug=False):
    data, pointer = read_file('day19.txt')
    run_code(data, pointer, debug=debug)

if __name__ == "__main__":
    main()


# Part 2 solved manually

File: 2018/Haskell/src/Day19.hs
{-# LANGUAGE DeriveFunctor #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE TemplateHaskell #-}

module Day19 where

import Control.Lens
import Data.Bits
import Data.Bool (bool)
import Data.Char
import Data.Functor.Foldable
import Data.Map (Map, (!))
import qualified Data.Map as Map
import Data.Sequence (Seq)
import qualified Data.Sequence as S
import Text.ParserCombinators.Parsec

type Memory = Map Int Int
data Opcode
  = Addr | Addi | Mulr | Muli
  | Banr | Bani | Borr | Bori
  | Setr | Seti | Gtir | Gtri
  | Gtrr | Eqir | Eqri | Eqrr
  deriving (Show, Eq, Ord, Read, Enum, Bounded)

data Instruction a = I
  { _opcode :: a
  , _a :: Int
  , _b :: Int
  , _c :: Int
  }
  deriving (Show, Eq, Ord, Functor)

makeLenses ''Instruction

data Program = P
  { _mem :: Memory
  , _ins :: Seq (Instruction Opcode)
  }
  deriving (Show, Eq, Ord)

makeLenses ''Program

runMemory :: Memory -> Instruction Opcode -> Memory
runMemory m (I op a b c) =
  case op of
    Addr -> Map.insert c (m ! a + m ! b) m
    Addi -> Map.insert c (m ! a + b) m
    Mulr -> Map.insert c (m ! a * m ! b) m
    Muli -> Map.insert c (m ! a * b) m
    Banr -> Map.insert c (m ! a .&. m ! b) m
    Bani -> Map.insert c (m ! a .&. b) m
    Borr -> Map.insert c (m ! a .|. m ! b) m
    Bori -> Map.insert c (m ! a .|. b) m
    Setr -> Map.insert c (m ! a) m
    Seti -> Map.insert c a m
    Gtir -> Map.insert c (bool 0 1 $ a > (m ! b)) m
    Gtri -> Map.insert c (bool 0 1 $ (m ! a) > b) m
    Gtrr -> Map.insert c (bool 0 1 $ (m ! a) > (m ! b)) m
    Eqir -> Map.insert c (bool 0 1 $ a == (m ! b)) m
    Eqri -> Map.insert c (bool 0 1 $ (m ! a) == b) m
    Eqrr -> Map.insert c (bool 0 1 $ (m ! a) == (m ! b)) m

eval :: Int -> Program -> Log Program
eval ipReg p@P{..} =
  if ip >= length _ins
    then Done $ _mem ! 0
    else Running $ p{_mem = runMemory _mem (S.index _ins ip)} & over (mem . ix ipReg) (+1)
 where
  ip = _mem ! ipReg

data Log a
  = Running a
  | Done Int
  deriving (Show, Eq, Ord, Functor)

runToHalt :: Log (Maybe Int) -> Maybe Int
runToHalt (Running a) = a
runToHalt (Done i) = Just i

part1 :: Int -> Program -> Maybe Int
part1 = (runToHalt `hylo`) . eval

main :: IO ()
main = do
  (ip, ops) <- parseInput <$> readFile "../data/day19.in"
  print $ part1 ip $ P (Map.fromList (zip [0 ..] (replicate 6 0))) (S.fromList ops)

-- | Parsing
parseInput :: String -> (Int, [Instruction Opcode])
parseInput =
  either (error . show) id
    . parse
      ( (,)
          <$> (read <$> (string "#ip " *> many1 digit <* newline))
          <*> sepEndBy1 parseInstr newline
      )
      ""
 where
  parseOpcode :: Parser Opcode
  parseOpcode = many1 alphaNum <&> read . over _head toUpper
  parseInstr :: Parser (Instruction Opcode)
  parseInstr =
    I <$> parseOpcode <* space
      <*> pNum <* space
      <*> pNum <* space
      <*> pNum
  pNum = read <$> many digit
